### PR TITLE
[FG:InPlacePodVerticalScaling] PLEG watch conditions: rapid polling for expected changes

### LINKED ISSUE #123940: [FG:InPlacePodVerticalScaling] PLEG doesn't work well with alpha feature InPlacePodVerticalScaling
#### Background

This is a follow up issue from https://github.com/kubernetes/kubernetes/pull/120432/files#r1489932247

Originally, we fix the InPlacePodVerticalScaling performance issue by fetching the runtime status in single sync loop which is not elegant. Later, I follow @smarterclayton's suggestion to leverage PLEG to emit events to fix it. 

https://github.com/kubernetes/kubernetes/blob/89f03e3988a4e7fed90ffce22f355ff248520ad2/pkg/kubelet/kubelet.go#L1988-L1995
(InPlacePodVerticalScaling puts the resizing pod into PLEG cache for further reconcilation)

I notice there're two problems.

#### 1. There's no way to generate a PodLifecycleEvent event for resized pod.

a.  oldPod and newPod are exact same in in-place update scenarios. It can not distinguish the pod that has been resized.  Since they are same, no events would be generated.

https://github.com/kubernetes/kubernetes/blob/89f03e3988a4e7fed90ffce22f355ff248520ad2/pkg/kubelet/pleg/generic.go#L252-L257C14

b. The running pod PLEG cache are not being used at all. 
`pleg.updateCache()` will call `runtime.GetPodStatus` underneath. In VPA case, the latest CRI container status will be fetched. That means the cache store a new container status. However, it is not being used in the `Relist` flow.  I think even we use it in`pleg.Relist()`, there's no way to distinguish the resized pod based on existing fields. Please check attached code snippets of their data structures.

#### 2. plegContainerUnknown(ContainerChanged) is not correctly handled for resized container

Seems `plegContainerUnknown(ContainerChanged)` is the best state for the resized container.  However, this is not correctly handle and it was not successfully passed in the event channel. 

https://github.com/kubernetes/kubernetes/blob/89f03e3988a4e7fed90ffce22f355ff248520ad2/pkg/kubelet/pleg/generic.go#L311-L312

#### Proposal
I would suggest to remove the `kl.pleg.UpdateCache(&runningPod, pod.UID);` logic in kubelet syncPod loop since the pod would be fetched in `pleg.Relist()` and properly handle the `ContainerChanged` event.

I cut a new PR to fix this issue https://github.com/kubernetes/kubernetes/pull/123941, please check whether that makes sense?

### What did you expect to happen?

I hope the resized container could be picked up by PLEG Relist logic (interval is 1 s) and trigger the status update and no need to wait for kubelet's next reconcile loop  (interval is 1 minute)

### How can we reproduce it (as minimally and precisely as possible)?

1. Enable the InPlacePodVerticalScaling feature gate and start the cluster
2. Create a pod with 1c.
3. Bump to 2c.
4. Check pod container status.  `ResourceAllocated` would be updated but status won't be updated until next kubelet recocnile loop, which means normally you will wait ~1 mins.

### Anything else we need to know?

relist pod status. The status is captured after the pod cpu was update from 1 to 2

old pod
```
{
  "ID": "4e2b3d81-129f-40e4-a579-631397aa718c",
  "Name": "tomcat",
  "Namespace": "default",
  "CreatedAt": 1710219686362299600,
  "Containers": [
    {
      "ID": "containerd://a06de85dbb87c2c0632df34b60ff9323ea29b9be362080ae75b7fdf0526e9c17",
      "Name": "tomcat",
      "Image": "sha256:ef6a7c98d192507d6066dcf24e44bec66d07ec9cf7c55d8d3d1ea0a24660bdef",
      "ImageID": "sha256:ef6a7c98d192507d6066dcf24e44bec66d07ec9cf7c55d8d3d1ea0a24660bdef",
      "ImageRef": "sha256:ef6a7c98d192507d6066dcf24e44bec66d07ec9cf7c55d8d3d1ea0a24660bdef",
      "ImageRuntimeHandler": "",
      "Hash": 2397588892,
      "HashWithoutResources": 3106650780,
      "State": "running"
    }
  ],
  "Sandboxes": [
    {
      "ID": "containerd://6bda38b639c84b369a169b94f8bf820bda49ab8f98a2ab365acc52883cebb25a",
      "Name": "",
      "Image": "",
      "ImageID": "",
      "ImageRef": "",
      "ImageRuntimeHandler": "",
      "Hash": 0,
      "HashWithoutResources": 0,
      "State": "running"
    }
  ]
}
```

newpod
```
{
  "ID": "4e2b3d81-129f-40e4-a579-631397aa718c",
  "Name": "tomcat",
  "Namespace": "default",
  "CreatedAt": 1710219686362299600,
  "Containers": [
    {
      "ID": "containerd://a06de85dbb87c2c0632df34b60ff9323ea29b9be362080ae75b7fdf0526e9c17",
      "Name": "tomcat",
      "Image": "sha256:ef6a7c98d192507d6066dcf24e44bec66d07ec9cf7c55d8d3d1ea0a24660bdef",
      "ImageID": "sha256:ef6a7c98d192507d6066dcf24e44bec66d07ec9cf7c55d8d3d1ea0a24660bdef",
      "ImageRef": "sha256:ef6a7c98d192507d6066dcf24e44bec66d07ec9cf7c55d8d3d1ea0a24660bdef",
      "ImageRuntimeHandler": "",
      "Hash": 2397588892,
      "HashWithoutResources": 3106650780,
      "State": "running"
    }
  ],
  "Sandboxes": [
    {
      "ID": "containerd://6bda38b639c84b369a169b94f8bf820bda49ab8f98a2ab365acc52883cebb25a",
      "Name": "",
      "Image": "",
      "ImageID": "",
      "ImageRef": "",
      "ImageRuntimeHandler": "",
      "Hash": 0,
      "HashWithoutResources": 0,
      "State": "running"
    }
  ]
}
```

pleg cached pod - from https://github.com/kubernetes/kubernetes/blob/89f03e3988a4e7fed90ffce22f355ff248520ad2/pkg/kubelet/kubelet.go#L1988-L1995

```
{
  "ID": "4e2b3d81-129f-40e4-a579-631397aa718c",
  "Name": "tomcat",
  "Namespace": "default",
  "IPs": [
    "10.88.0.63",
    "2001:db8:4860::3f"
  ],
  "ContainerStatuses": [
    {
      "ID": "containerd://a06de85dbb87c2c0632df34b60ff9323ea29b9be362080ae75b7fdf0526e9c17",
      "Name": "tomcat",
      "State": "running",
      "CreatedAt": "2024-03-12T05:01:27.032584855Z",
      "StartedAt": "2024-03-12T05:01:27.083899049Z",
      "FinishedAt": "0001-01-01T00:00:00Z",
      "ExitCode": 0,
      "Image": "docker.io/library/tomcat:8.0",
      "ImageID": "docker.io/library/tomcat@sha256:8ecb10948deb32c34aeadf7bf95d12a93fbd3527911fa629c1a3e7823b89ce6f",
      "ImageRef": "docker.io/library/tomcat@sha256:8ecb10948deb32c34aeadf7bf95d12a93fbd3527911fa629c1a3e7823b89ce6f",
      "ImageRuntimeHandler": "",
      "Hash": 2397588892,
      "HashWithoutResources": 3106650780,
      "RestartCount": 0,
      "Reason": "",
      "Message": "",
      "Resources": {
        "CPURequest": "2",
        "CPULimit": "2",
        "MemoryRequest": null,
        "MemoryLimit": null
      }
    }
  ],
  "SandboxStatuses": [
    {
      "id": "6bda38b639c84b369a169b94f8bf820bda49ab8f98a2ab365acc52883cebb25a",
      "metadata": {
        "name": "tomcat",
        "uid": "4e2b3d81-129f-40e4-a579-631397aa718c",
        "namespace": "default"
      },
      "created_at": 1710219686362299600,
      "network": {
        "ip": "10.88.0.63",
        "additional_ips": [
          {
            "ip": "2001:db8:4860::3f"
          }
        ]
      },
      "linux": {
        "namespaces": {
          "options": {
            "pid": 1
          }
        }
      },
      "labels": {
        "io.kubernetes.pod.name": "tomcat",
        "io.kubernetes.pod.namespace": "default",
        "io.kubernetes.pod.uid": "4e2b3d81-129f-40e4-a579-631397aa718c"
      },
      "annotations": {
        "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"v1\",\"kind\":\"Pod\",\"metadata\":{\"annotations\":{},\"name\":\"tomcat\",\"namespace\":\"default\"},\"spec\":{\"containers\":[{\"image\":\"tomcat:8.0\",\"imagePullPolicy\":\"Always\",\"name\":\"tomcat\",\"ports\":[{\"containerPort\":7500}],\"resizePolicy\":[{\"resourceName\":\"cpu\",\"restartPolicy\":\"NotRequired\"}],\"resources\":{\"limits\":{\"cpu\":1},\"requests\":{\"cpu\":1}}}]}}\n",
        "kubernetes.io/config.seen": "2024-03-12T05:01:26.046824676Z",
        "kubernetes.io/config.source": "api"
      }
    }
  ],
  "TimeStamp": "0001-01-01T00:00:00Z"
}
```

### Kubernetes version

master version

### Cloud provider

Common problem so it applies to any cloud providers.

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by tallclair**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

**Motivation & Background:**

The Kubelet's PLEG only triggers events and fetches the latest PodStatus from the runtime when a container changes state. This is a problem for in-place pod resize, since resource updates are not considered a state change. To get around this, `pleg.UpdateCache` was called at the end of SyncPod if a pod was resizing to force a "reinspection" of the pod status. This violates some principles of the PLEG, and also means that resizes need to wait a full resync period (1 minute) to detect that the resize completed.

**What this PR does:**

This PR adds a new concept of "Watch Conditions" to the PLEG. If a pod has a watch condition set, then the pod is reinspected (via `runtime.GetPodStatus`) on every `Relist` loop (every 2 seconds) until the watch condition function returns true, indicating that the condition was reached. The Kubelet sets a watch condition for each resource resize, so it no longer needs to manually do the reinspection via update cache.

Not Yet Implemented: When a watch condition is completed, an event is emitted by the PLEG to trigger SyncPod. This should greatly speed up pod resize completion.

#### Which issue(s) this PR fixes:
Fixes #123940

#### Special notes for your reviewer:
This PR starts with some refactoring. I can pull the refactor into a separate PR if you prefer.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/sig node
/priority important-longterm

### CODE REVIEW

**Comment by tallclair** [comment]
I pulled in the change to block on the ResizeStatus being cleared from https://github.com/kubernetes/kubernetes/pull/128377, and ran the e2e tests locally. Without this PR, the tests timed out (at 27/35 resize tests). With this PR, the tests complete in about 10 minutes.

**Comment by tallclair** [comment]
Rebased.

File: pkg/kubelet/pleg/generic.go [hunk 1]
```diff
@@ -250,27 +250,26 @@ func (g *GenericPLEG) Relist() {
...
+		// Compare the old and the current pods, and generate events.
 		oldPod := g.podRecords.getOld(pid)
 		pod := g.podRecords.getCurrent(pid)
 		// Get all containers in the old and the new pod.
 		allContainers := getContainersFromPods(oldPod, pod)
+		var events []*PodLifecycleEvent
 		for _, container := range allContainers {
-			events := computeEvents(g.logger, oldPod, pod, &container.ID)
-			for _, e := range events {
-				updateEvents(eventsByPodID, e)
```
Comment (yujuhong) [inline pkg/kubelet/pleg/generic.go:263]: Delete the `updateEvents` function since this is the only call site

**Review by yujuhong — COMMENTED** [review]
The refactor looks good. It simplified the code now that we don't have to condition on the existence of the cache. 

I still need to review the main (watch condition) changes, but from a quick glance, I think it could use some more comments to explain the condition key, versioning, and the necessity of those.

File: pkg/kubelet/pleg/generic.go (hunk 1 above)
Comment (tallclair) [reply pkg/kubelet/pleg/generic.go:263]: Looks like that change slipped into the next commit.

**Comment by tallclair** [comment]
Thanks, I'll try to proactively add some explanatory comments.

File: pkg/kubelet/pleg/generic_test.go [hunk 2]
```diff
@@ -736,3 +736,155 @@ kubelet_running_pods 2
 		})
 	}
 }
+
+func TestWatchConditions(t *testing.T) {
```
Comment (yujuhong) [inline pkg/kubelet/pleg/generic_test.go:740]: It seems like we could add more tests, e.g.,
 - Multiple containers in the pod
 - Condition not cleared when evaluating an older version

File: pkg/kubelet/pleg/generic.go [hunk 3]
```diff
@@ -460,22 +480,93 @@ func (g *GenericPLEG) updateCache(ctx context.Context, pod *kubecontainer.Pod, p
...
+		}
+		conditions = make(map[string]versionedWatchCondition)
+	}
+
+	versioned, found := conditions[conditionKey]
+	if found {
+		versioned.version++
+		versioned.condition = condition
+		conditions[conditionKey] = versioned
+	} else if condition != nil {
```
[the remark this answers is not in the document]
Comment (tallclair) [reply pkg/kubelet/pleg/generic.go]: This code got left behind from an earlier version where passing a `nil` condition was used to clear the condition. We can add it back if we need that capability, but I've removed it for now.

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 4]
```diff
@@ -797,6 +798,22 @@ func (m *kubeGenericRuntimeManager) updatePodContainerResources(pod *v1.Pod, res
 				"pod", format.Pod(pod), "resourceName", resourceName)
 			return err
 		}
+		resizeKey := fmt.Sprintf("%s:resize:%s", container.Name, resourceName)
```
[the remark this answers is not in the document]
Comment (tallclair) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go:801]: This function will get called on every SyncPod iteration until the resources converge. I think that's OK here. Updating the resources should take much less than the 1 minute resync period, so I'd expect this to be a very small number of times updating the watch condition. In the restart case, the watch condition will get added back if the resources still haven't converged.

**Comment by knabben** [comment]
Hey @yujuhong @tallclair 
⚠️ Do we still intend to merge this for v1.32? Just a reminder that the code freeze is starting 02:00 UTC Friday November 8th 2024 (a little less than 1 week from now). Please make sure the PR has both lgtm and approved labels before the code freeze. Thanks!

File: pkg/kubelet/pleg/generic_test.go (hunk 2 above)
Comment (tallclair) [reply pkg/kubelet/pleg/generic_test.go:742]: Done.

**Comment by yujuhong** [comment]
/lgtm

Can we also run the resize test to verify?

**Comment by tallclair** [comment]
Fixed lint error

**Pull request merged**